### ISSUE TITLE
Document Subscription

### ISSUE THREAD

**Issue by MrVauxs** [foundryvtt/foundryvtt#11980]
### User Experience

It would be great to be able to subscribe to documents and listen to any changes being made to them, or at the very least be notified that a document has changed without the need to listen to a hook. There are plenty implementations to take inspiration from such as Svelte [Runes](https://svelte.dev/docs/svelte/what-are-runes)/[Stores](https://svelte.dev/docs/svelte/svelte-store), [Solid](https://www.solidjs.com/tutorial/introduction_signals?solved), [the official Signals proposal](https://github.com/tc39/proposal-signals) etc.

This would allow for fine-grained reactivity without the need for document wrappers or Hooks that trigger at overbroad triggers.

```js
const actor = game.actors.get("...")

actor.subscribe((document) => {
    // Do something reactively to this specific actor changing
})
```

**Comment by JPMeehan** [comment]
It's not clear to me what you're getting here that's different from the existing hooks framework.

**Comment by MrVauxs** [comment]
> It's not clear to me what you're getting here that's different from the existing hooks framework.

Hooks trigger on any document being updated, requiring filtering down and storing what you are tracking to be able to have impromptu reactivity. As said in the proposal, they are too wide of a net, this is a suggestion to be able to be fine-grained.
You also have to have 2 Hooks to account for document deletion. 

Also, to quote @typhonrt in Discord (and who could also chime in with what this would entail)
> Currently, to get callbacks from the document model for not app constructs you have to fake the shape of Application providing an object w/ `render` (document updated) & `close` (document deleted) functions.

**Comment by MrVauxs** [comment]
The Pathfinder 2e system is also starting to embrace Svelte into its applications, but due to the nature of documents, the way reactivity works is through passing the entire document JSON object on `render` calls, which is hardly efficient.

**Comment by BoltsJ** [comment]
What's stopping systems that need this from adding the necessary interface to their document classes and putting the handling in _onUpdate and _onDelete if hooks are too broad? Why would it need to be in core foundry?

**Comment by MrVauxs** [comment]
Intercompatibility and having one unified interface.

Significantly nicer dev experience, not having to monkeypatch very
important methods from the system side.

It would be a major improvement to the ecosystem, it would streamline the
creation of Apps ***significantly*** without reliance on absolute `render`
methods.

niedz., 29 gru 2024, 23:06 użytkownik Joe ***@***.***>
napisał:

**Comment by MrVauxs** [comment]
And being able to subscribe only whenever a specific value changes makes
for an efficient and much dev friendlier UI solution instead of
re-rendering the entire innerHTML every time something in the document
changed.

niedz., 29 gru 2024, 23:13 użytkownik MrVauxs ***@***.***> napisał:

**Comment by aaclayton** [comment]
Wasn't in scope for V13 but I think in V14 the `DataModel` will have an event emitter pattern added to it so you can register listeners on a data model to respond to specific events for that particular model.

**Comment by MrVauxs** [comment]
> Wasn't in scope for V13 but I think in V14 the `DataModel` will have an event emitter pattern added to it so you can register listeners on a data model to respond to specific events for that particular model.

Just to elaborate, would these subscriptions be for the specific documents or for the broader DataModel classes?

**Comment by aaclayton** [comment]
They would be for specific data model instances, similar to how the new event emitter design works for ApplicationV2 instances.

**Comment by aaclayton** [comment]
> Is adding this event pattern still in V14's scope?

Not currently, but I am enthusiastic about the idea.

**Comment by MrVauxs** [comment]
What is the status on this issue right now?

**Comment by aaclayton** [comment]
> What is the status on this issue right now?

V15 maybe. Not really an architectural priority in my mind. Nice to have, but depends on V15 scope whether it fits or not.

**Comment by typhonrt** [comment]
Once again I'd be glad to provide some insight on how this can be implemented and perhaps provide a proof of concept. This is really the key to unlocking _any_ reactive framework. The hack currently being used which is the client document / app hard associated callbacks sort of works. This basic mechanism has been around forever. 

There are many plain JS reasons to have a solid API here regardless of reactive frameworks. IE it's a widely accessible / useful feature. I firmly believe that improving this area will also help the AppV2 infrastructure significantly.

**Comment by typhonrt** [comment]
One thing I'd like to add is that the proposed V2 observability API provides a broader set of use cases beyond Application synchronization enabling  a more `data-first` paradigm. Rather than documents primarily notifying Application instances via `ClientDocumentMixin`, document state itself becomes observable and this opens the door for any consumer to react to changes through a streamlined common API.

The current limitation is that the `Application` observation mechanism isn't really a general document observation API. It's tightly coupled to the Application lifecycle, assumes methods like `render()` and `close()`, and lacks explicit subscribe / unsubscribe semantics. That's why I see a potential new subscriber API as a complementary addition rather than a replacement. It shifts observability to being  a _document model primitive_ instead of an application specific feature only available on the client. Another significant potential win is even server-side document instances having observability built-in for future use cases on the backend.

I'm going to try and free up some time in the coming weeks to make a trial local implementation and subsequent video to post demoing anything I discover. The primary goal is a very minimal change to the document model API surface that is opt-in for future use cases.

**Comment by aaclayton** [comment]
This is an issue of _prioritization_, not one of _difficulty_ or clarity of _use case_. We understand there is a use case and we understand it's not a huge lift to add an event emitter to the Document, but doing it properly is a significant enough task for a very niche use case. The reason we haven't done this is simply an output of our internal prioritization process.

And before you jump in and try and convince me that the use case is not niche, let me interrupt any such attempt by asserting that it is niche.

There is no chance of us doing this before V15.

**Comment by typhonrt** [comment]
Thanks for the clarification. I understand the prioritization concerns given the amount of ongoing work across Foundry from new core features to additional product work such as Ember, Crucible, and other initiatives.

I also understand the hesitation around introducing a new API that won’t directly be exercised by core workflows / core team in current efforts. While any potential document-local observation API may be a longer-term consideration, I do believe I can help reduce uncertainty around what a minimal and future oriented solution might look like. I'll experiment with a local prototype and document any findings around API shape, lifecycle concerns, and implementation tradeoffs.

Besides a document local push based solution there are interesting async pull based avenues to explore that go hand-in-hand and I do believe any successful demo will aid in future architectural improvements that broadly serve the larger development community.

I do have fish to fry so to speak on my side as well to catch up, so it may be a minute before any demo appears.

**Comment by MarkPearce** [comment]
Not Niche if you make modules.
